micro-rdk/src/common/conn/viam.rs

```diff
@@ -734,35 +774,58 @@ where
         mut app_client: Option<AppClient>,
     ) -> Result<(), errors::ServerError> {
         let wait = Duration::from_secs(1); // should do exponential back off
+
         loop {
-            if let Some(app_client) = app_client {
-                let mut app_client_tasks: FuturesUnordered<AppClientTaskRunner> =
-                    FuturesUnordered::new();
+            if let Some(app_client) = &app_client {
+                log::info!("starting execution of app client tasks");
+                let mut app_client_tasks: FuturesUnordered<_> = FuturesUnordered::new();
                 for task in &self.app_client_tasks {
-                    app_client_tasks.push(AppClientTaskRunner {
-                        app_client: &app_client,
-                        invoker: task,
-                        state: TaskRunnerState::Run {
-                            task: task.invoke(&app_client),
-                        },
+                    app_client_tasks.push({
+                        async move {
+                            let mut duration = task.get_default_period();
+                            loop {
+                                let _ = Timer::after(duration).await;
+                                match task.invoke(app_client).await {
+                                    Ok(Some(next_duration)) => {
+                                        duration = next_duration;
+                                    }
+                                    Ok(None) => {}
+                                    e @ Err(_) => {
+                                        break e;
+                                    }
+                                }
+                            }
+                        }
                     });
                 }
                 while let Some(res) = app_client_tasks.next().await {
                     if let Err(err) = res {
-                        log::error!("a task returned the following error {:?}", err);
+                        log::error!(
+                            "an app client task returned an error {:?} - dropping app client",
+                            err
+                        );
                         break;
                     }
                 }
             }
+
+            // Explicitly release the app_client resources before sleeping, and before
```

Comment:
I don't know if this is actually necessary. I need to better familiarize myself with the lifetime rules in rust, but the theory is that the defunct `AppClient` wouldn't get destroyed until after the timeout and the `self.connect_to_app().await` had completed and returned an object to overwrite it. But there is no point holding on to the dead `AppClient` while we build the new one. If I'm wrong about that and `Rust` will drop it before calling `connect_to_app`, then this can go away.

Reply:
So ok to explicitly drop it here although drop will be repeatably called on None until a connection is established.